### PR TITLE
Allow deployments from UI and deprecate rest2tasks

### LINKED ISSUE #838: ability to trigger deployment of an environment via the UI

### PR DESCRIPTION
**Pull request by rocketeerbkw**
# Checklist
- [x] Affected Issues have been mentioned in the Closing issues section
- [ ] Documentation has been written/updated.
- [x] Changelog entry has been written

### API
New mutations added to the API to trigger deployments. Examples:

```graphql
# Deploy latest code for an environment that already exists
mutation deployEnvironmentLatest {
  deployEnvironmentLatest (input: {
    environment: {
      name: "nginx"
      project: {
        name: "ci-nginx"
      }
    }
  })
}

# Deploy latest or specific commit for an environment that does or does not already exist
mutation deployEnvironmentBranch {
  deployEnvironmentBranch(input:{
    project: {
      name: "ci-nginx"
    }
    branchName: "nginx"
    branchRef: "b82sx9234"
  })
}

mutation deployEnvironmentPullrequest {
  deployEnvironmentPullrequest(input: {
    project: {
      name: "ci-nginx"
    }
    number: 5
    title: "111 Add New Feature"
    baseBranchName: "master"
    baseBranchRef: "origin/master"
    headBranchName: "111-add-new-feature"
    headBranchRef: "origin/111-add-new-feature"
  })
}

mutation deployEnvironmentPromote {
  deployEnvironmentPromote(input: {
    sourceEnvironment: {
      name: "source"
      project: {
        name: "ci-features"
      }
    }
    branchName:"target"
    project: {
      name:"ci-features"
    }
  })
}
```

In order to facilitate deployments, new information is tracked for environments and is now required when creating/updating them. Example:

```json
"environmentByOpenshiftProjectName": {
  "id": 3,
  "openshiftProjectName": "high-cotton-master",
  "name": "Master",
  "deployType": "branch",
  "deployTitle": null,
  "deployBaseRef": "Master",
  "deployHeadRef": null
}
```

### UI

Environments that have been configured with deployment information will have a new button on the deployments page to trigger a deploy of the latest code for that environment.

image

# Changelog Entry

Feature - API has new mutations for triggering deployments
Change - API requires new deployment information when creating/updating environments
Improvement - UI can now trigger new deployments for environments (#838)
Change - rest2tasks is deprecated in favor of API

## Deployment notes
For existing environments, at lest one deployment must happen via non-ui methods before the "deploy" button will show up in the ui. New environments will get the button immediately.

# Closing issues
Closes #838

### CODE REVIEW

File: services/api/src/typeDefs.js [hunk 1]
```diff
@@ -847,6 +878,32 @@ const typeDefs = gql`
...
+    baseBranchName: String!
+    baseBranchRef: String!
+    headBranchName: String!
+    headBranchRef: String!
+  }
+
+  input DeployEnvironmentPromoteInput {
+    sourceEnvironment: EnvironmentInput!
+    project: ProjectInput!
+    branchName: String!
```
Comment (Schnitzel) [inline services/api/src/typeDefs.js]: let's call this `destinationEnvironment` and not `branchName`. Calling it `branchName` was a weird idea when I created the Promote System.

The idea is basically that with a promotion you can take any Docker Images and Services from one Environment and create a second Environment.

so it makes more sense to me to have a `sourceEnvironment` and a `destinationEnvironment` defined

**Review by Schnitzel — CHANGES_REQUESTED** [review]

**Comment by Schnitzel** [comment]
very nice work! just some small adaptions and unfortunately we have some conflicts now :(

File: services/api/src/typeDefs.js (hunk 1 above)
Comment (rocketeerbkw) [reply services/api/src/typeDefs.js]: Done!

**Comment by rocketeerbkw** [comment]
I rebased to fix the conflicts, will wait for tests to come back while I test locally.

**Comment by rocketeerbkw** [comment]
All green for me locally

**Comment by Schnitzel** [comment]
ugh, another Conflict now :( @rocketeerbkw can you have another look? thank you

**Review by Schnitzel — APPROVED** [review]

**Pull request merged**